projects/dev/src/app/group/group-manage/pending-request/pending-request.component.ts

```diff
@@ -22,87 +37,149 @@ export class PendingRequestComponent implements OnInit, OnChanges {
-        } (${req.joining_user.login || ""})`,
-        "grade": req.joining_user && req.joining_user.grade ? req.joining_user.grade : null,
-        at: req.at,
+  _handleActionResponse(result: RequestActionResponse, verb: string, msg: string) {
+    if (result.success === true && result.message === "updated") {
+      const succ = _.countBy(result.data, (status: string) => {
+        return ["success", "unchanged"].includes(status);
       });
+
+      if (succ.false === undefined || succ.false === 0) {
+        this.messageService.add({
+          severity: "success",
+          summary: "Success",
+          detail: `${succ.true} request(s) have been ${msg}`,
+          life: TOAST_LENGTH,
+        });
+      } else if (succ.true === undefined || succ.true === 0) {
+        this.messageService.add({
+          severity: "error",
+          summary: "Error",
+          detail: `Unable to ${verb} the selected request(s).`,
+          life: TOAST_LENGTH,
+        });
+      } else {
+        this.messageService.add({
+          severity: "warn",
+          summary: "Partial success",
+          detail: `${succ.true} request(s) have been ${msg}, ${succ.false} could not be executed`,
+          life: TOAST_LENGTH,
+        });
+      }
+
+      this.reloadData();
     }
   }
 
-  constructor(private groupService: GroupService) {}
+  _processRequestError(_err) {
+    this.messageService.add({
+      severity: "error",
+      summary: "Error",
+      detail: ERROR_MESSAGE.fail,
+      life: TOAST_LENGTH,
+    });
+  }
+
+  constructor(
+    private groupService: GroupService,
+    private messageService: MessageService
+  ) {}
 
   ngOnInit() {
     this.panel.push({
       name: "Pending Requests",
       columns: this.columns,
     });
-    this.groupService
-      .getManagedRequests(this.id)
-      .subscribe((reqs: PendingRequest[]) => {
-        this._setRequestData(reqs);
-      });
   }
 
-  ngOnChanges(changes: SimpleChanges) {
-    this.groupService
-      .getManagedRequests(this.id)
-      .subscribe((reqs: PendingRequest[]) => {
-        this._setRequestData(reqs);
-      });
+  ngOnChanges(_changes: SimpleChanges) {
+    this.reloadData();
   }
 
-  onExpandWidth(e) {}
+  onExpandWidth(_e) {}
 
-  onClickAccept(e) {
-    this.onAccept.emit(e);
+  onProcessRequest(type: string) {
+    if (this.selection.length === 0 || this.requestLoading) {
+      return;
+    }
+
+    let resultObserver: Observable<RequestActionResponse>;
+    this.requestLoading = true;
+    this.requestAction = type;
+    
+    const group_ids = this.selection.map((req: PendingRequest) => req.joining_user.group_id);
+
+    if (type === "accept") {
+      resultObserver = this.groupService.acceptJoinRequest(this.id, group_ids);
+    } else {
+      resultObserver = this.groupService.rejectJoinRequest(this.id, group_ids);
+    }
+
+    resultObserver.subscribe(
+      (res: RequestActionResponse) => {
+        this._handleActionResponse(
```

Comment:
The separation of responsibility between `_handleActionResponse` and  `onProcessRequest` is not clear.  Why for instance reloading in `_handleActionResponse` and not here? Why setting `this.requestLoading = false` here and not at the other place?
I don't say that to annoy you but if you do not have a clear separation, it is hard to check that a behavior is correct and it explains why you encounter bugs.

Reply:
The `requestLoading` variable is not here anymore.
Instead, I use `requestAction` variable and validate if the action is in progress with `None`.
The reason why the setting should be here is that we should wait for the accept/reject api request finishes.
Before that, we don't have to set the variable to active so that the accept/reject button being clicked again which will call the api endpoint again while the first request is not finished yet.

Reply:
Both `_handleActionResponse`call and `this.requestAction = AcceptReject.None; ... ` run when the request when the response has been received, so it does not answer my concern. 
So why putting `this._reloadData();` in `_handleActionResponse()` and not after `this.requestAction = AcceptReject.None;`. Why not putting `this.requestAction = AcceptReject.None;` in `_handleActionResponse(..)` instead of where it is? You need to know why it is there, the code cannot be organized randomly. I do not say it is wrong, I just want to know.